File: python/twli_consolidate/crosswalk.py
"""Load and represent village consolidation events from the YAML crosswalk."""
from __future__ import annotations

from dataclasses import dataclass, field
from importlib import resources
from pathlib import Path
from typing import Iterable

import yaml


VALID_TYPES = {"rename", "merge", "split", "redistribute", "boundary_adjust"}


def bundled_crosswalk_path() -> Path:
    """Return the file path of the crosswalk YAML shipped with the package."""
    with resources.as_file(resources.files(__package__) / "data" / "village_changes.yaml") as p:
        return Path(p)


@dataclass(frozen=True)
class Event:
    """One administrative-boundary change event between two annual snapshots."""
    id: str
    effective_year: int          # Gregorian year from which the new boundary applies
    type: str                    # one of VALID_TYPES
    sources: tuple[str, ...]     # V_IDs in the prior snapshot
    targets: tuple[str, ...]     # V_IDs in the new snapshot
    raw: dict = field(repr=False, default_factory=dict)

    def affected_vids(self) -> set[str]:
        return set(self.sources) | set(self.targets)


def load_crosswalk(path: str | Path) -> list[Event]:
    """Load the crosswalk YAML file and return a list of Event objects."""
    path = Path(path)
    with path.open(encoding="utf-8") as f:
        doc = yaml.safe_load(f)
    raw_events = doc.get("events", [])
    out: list[Event] = []
    for ev in raw_events:
        etype = ev["type"]
        if etype not in VALID_TYPES:
            raise ValueError(f"Unknown event type {etype!r} in event {ev.get('id')}")
        srcs = tuple(item["v_id"] for item in ev.get("sources", []))
        tgts = tuple(item["v_id"] for item in ev.get("targets", []))
        out.append(Event(
            id=ev["id"],
            effective_year=int(ev["effective_year"]),
            type=etype,
            sources=srcs,
            targets=tgts,
            raw=ev,
        ))
    return out


def filter_by_year_range(events: Iterable[Event], year_range: tuple[int, int]) -> list[Event]:
    """Keep only events whose effective_year falls inside (start_year, end_year].

    An event with effective_year = Y means: starting from year Y the new boundary
    is in effect. If the user's panel covers years [start, end], only events
    with start < Y <= end actually cause a boundary change inside the panel.
    """
    start, end = year_range
    return [e for e in events if start < e.effective_year <= end]

File: R/twliConsolidate/R/crosswalk.R
#' Valid event types in the village consolidation crosswalk.
#' @keywords internal
.VALID_EVENT_TYPES <- c("rename", "merge", "split", "redistribute",
                        "boundary_adjust")

#' Load the village consolidation crosswalk YAML.
#'
#' @param path Path to the YAML file (e.g. ``"crosswalk/village_changes.yaml"``).
#' @return A list of event records. Each event is a list with elements
#'   ``id``, ``effective_year`` (Gregorian integer), ``type``,
#'   ``sources`` (character vector of V_IDs in the prior snapshot),
#'   ``targets`` (character vector of V_IDs in the new snapshot), and
#'   ``raw`` (the original YAML record).
#' @export
load_crosswalk <- function(path) {
  doc <- yaml::read_yaml(path)
  raw_events <- doc$events %||% list()
  lapply(raw_events, function(ev) {
    if (!ev$type %in% .VALID_EVENT_TYPES) {
      stop(sprintf("Unknown event type %s in event %s", ev$type, ev$id))
    }
    list(
      id              = ev$id,
      effective_year  = as.integer(ev$effective_year),
      type            = ev$type,
      sources         = vapply(ev$sources %||% list(),
                               function(x) x$v_id, character(1)),
      targets         = vapply(ev$targets %||% list(),
                               function(x) x$v_id, character(1)),
      raw             = ev
    )
  })
}

#' Restrict an event list to those whose effective_year falls inside
#' a given panel range.
#'
#' Events with ``start_year < effective_year <= end_year`` are kept.
#'
#' @param events A list of events from [load_crosswalk()].
#' @param year_range Length-2 integer vector ``c(start, end)`` (Gregorian).
#' @export
filter_by_year_range <- function(events, year_range) {
  stopifnot(length(year_range) == 2L)
  start <- year_range[[1]]; end <- year_range[[2]]
  Filter(function(e) e$effective_year > start && e$effective_year <= end, events)
}

`%||%` <- function(a, b) if (is.null(a)) b else a
